src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -614,6 +622,9 @@ public bool Remove(TKey key)
                         {
                             entries[last].next = entries[i].next;
                         }
+
+                        value = entries[i].value;
```

Comment:
> Are we referring to maintenance cost? But then what is special in context of generic type...

Maintenance cost too since we'll end up with 2 very similar pieces of code. But this is a generic type, this means that if you duplicate that code you're not adding just a single method, you're adding many more - one for each `Dictionary` instantiation where `TValue` is a value type (e.g. one for `Dictionary<int, int>`, another one for `Dictionary<int, bool>` and so on).

> In non-duplication way, I'm not able to understand how we could avoid assignment at line 626...

I don't know exactly what @danmosemsft had in mind when he said branch but one possible option is to pass a `bool` parameter to the method indicating whether to perform the copy or not. That would imply moving all this code to a new private method and having both `Remove` overloads call that method.

> I was curious why the C# compiler would not inline entire Remove(key, out value) into the...

The C# compiler doesn't do inlining, that's the job of the JIT compiler since inlining normally needs to take into account various "low level" aspects (e.g. native code size, calling convention). Anyway, there's too much code to inline.

> Performance data points I will collect

Make sure to measure the case of large value types. I don't expect this change to have a measurable impact on `Dictionary<int, int>` but the story may be different for something like `Dictionary<int, Matrix>`.